### PR TITLE
Tobin gh91 new cb invalid cases

### PR DESCRIPTION
**Pull request by tobine**
Reworking of image and buffer binding to cmd buffer.
Now correctly handle case where cmd buffer image/buffer dependencies that are destroyed will cause cmd buffer to become INVALID w/ and error flagged at submit time. Added tests for image/buffer cases.

Still more cases like this to add as well as their tests, but this is a good start and will address #641

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -416,6 +416,16 @@ static bool validate_buffer_usage_flags(layer_data *dev_data, VkBuffer buffer, V
     return skip_call;
 }
 
+// Helper function to validate usage flags for buffers
+// For given buffer_node send actual vs. desired usage off to helper above where
+//  an error will be flagged if usage is not correct
+static bool validateBufferUsageFlags(layer_data *dev_data, BUFFER_NODE const *buffer_node, VkFlags desired, VkBool32 strict,
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Can this completely replace `validate_buffer_usage_flags` just above?

Comment (tobine) [inline layers/core_validation.cpp]: Yes, pushed. I'll rework this with #686 when they collide.

File: tests/layer_validation_tests.cpp [hunk 2]
```diff
@@ -5267,6 +5267,83 @@ TEST_F(VkLayerTest, WriteDescriptorSetIntegrityCheck) {
     vkDestroyDescriptorPool(m_device->device(), ds_pool, NULL);
 }
 
+TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
+    TEST_DESCRIPTION("Attempt to draw with a command buffer that is invalid "
+                     "due to an image dependency being destroyed.");
+    ASSERT_NO_FATAL_FAILURE(InitState());
+
+    VkImage image;
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: This test seems far more complex than it needs to be.

Comment (tobine) [inline tests/layer_validation_tests.cpp]: Yeah, I thought I'd be clever and use "vkCmdCopyBufferToImage()" so I could share code between buffer/image tests, but I agree it's too complex for what it's doing.

Pushed a simpler version that just uses vkCmdClearColorImage()

File: tests/layer_validation_tests.cpp [hunk 3]
```diff
@@ -5819,40 +5819,20 @@ TEST_F(VkLayerTest, InvalidCmdBufferImageDestroyed) {
...
-                           image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1,
-                           &region);
+    VkClearColorValue ccv;
+    ccv.float32[0] = 1.0f;
+    ccv.float32[1] = 1.0f;
+    ccv.float32[2] = 1.0f;
+    ccv.float32[3] = 1.0f;
+    VkImageSubresourceRange isr = {};
+    isr.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+    isr.baseArrayLayer = 1;
```
Comment (chrisforbes) [inline tests/layer_validation_tests.cpp]: baseArrayLayer and baseMiplevel should be 0.

**Comment by chrisforbes** [comment]
With the test fixed, :+1:

**Comment by tobine** [comment]
cherry-picked and pushed to master

**Pull request closed without merging**